### PR TITLE
Refactor dependencies.yaml to use depends-on pattern.

### PR DESCRIPTION
**Pull request by bdice**
This refactors cuvs's `dependencies.yaml` to follow a convention we've established in other repos. For RAPIDS packages we declare a `depends_on_X` dependency list, and use that rather than specifying packages multiple times. This enables more reuse and easier automation for version migrations.

### CODE REVIEW

File: python/cuvs/pyproject.toml [hunk 1]
```diff
@@ -38,6 +38,7 @@ dependencies = [
     "nvidia-cusolver",
     "nvidia-cusparse",
     "pylibraft==24.10.*,>=0.0.0a0",
+    "rmm==24.10.*,>=0.0.0a0",
```
Comment (jameslamb) [inline python/cuvs/pyproject.toml:41]:
```suggestion
```

I'm not sure about this... it seems to me that `cuvs` doesn't have a direct runtime dependency on the `rmm` Python package.

```shell
git grep -i rmm python/
```

If this wasn't intentional, it should be removed.

**Review by jameslamb — APPROVED** [review]
I love this pattern generally. Left one question for your consideration, I don't need to review again after you've addressed it.

File: python/cuvs/pyproject.toml (hunk 1 above)
Comment (bdice) [reply python/cuvs/pyproject.toml:41]: Agreed. I'll clean this up.

Comment (bdice) [reply python/cuvs/pyproject.toml]: I ended up doing a partial cleanup in this PR and a partial cleanup in #322, to avoid merge conflicts.

Comment (jameslamb) [reply python/cuvs/pyproject.toml]: it looks great, thank you for that

**Review by jameslamb — APPROVED** [review]
re-approving. Really glad we could cut some unnecessary build dependencies here!

**Pull request merged**